src/document/Document.js

```diff
@@ -122,6 +125,217 @@ export const Document = class {
+      } else {
+        return null;
+      }
+    }
+
+    const collect = (node) => {
+      const kept = !!node.field;
+      const keep = [];
+      for (const child of node.childNodes) {
+        keep.push(...collect(child));
+      }
+
+      if (kept) {
+        return [node];
+      } else {
+        return keep;
+      }
+    }
+
+    const addAll = (node) => {
+      // collect descendants to include
+      const include = [];
+      let current = collect(node);
+      while (current.length > 0) {
+        include.push(...current);
+        let children = []
+        for (n of current) {
+          children.push(...n.childNodes.map(n => collect(n)));
+        }
+        current = children;
+      }
+
+      const obj = {};
+      for (const el of include) {
+        const value = el.toString();
+        const fields = [el.field, ...el.extraFields];
+        for (const field of fields) {
+          // Add value(s) to current object
+          if (field in obj) {
+            if (!Array.isArray(obj[field])) {
+              obj[field] = [obj[field]];
+            }
+            obj[field].push(value);
+          } else {
+            obj[field] = value;
+          }
+        }
+      }
+      return obj;
+    }
+
+    const toObj = (include) => {
+      const result = [];
+      let obj = {};
+      let context = {};
+      let prev = null;
+
+      for (const el of include) {
+        const fields = [el.field, ...el.extraFields];
+        for (const field of fields) {
+          if (fields.includes('_container')) {
+            result.push(addAll(el));
+            continue;
+          } else {
+            // Otherwise, treat as sequential
+            const value = el.toString();
+
+            // Handle shared fields (e.g., generation as a header)
+            if (sharedFields.includes(field)) {
+              if (Object.keys(obj).length != 0) {
+                result.push({ ...context, ...obj });
+              }
+              obj = {};
+              context[field] = value;
+              prev = null;
+              continue; // Wait for a normal field to add to the object
+            }
+
+            // Check if this field starts a new object
+            if (normalFields.includes(field) && field in obj && prev != field) {
+              result.push({ ...context, ...obj });
+              obj = {};
+            }
+
+            // Add value(s) to current object
+            if (field in obj) {
+              if (!Array.isArray(obj[field])) {
+                obj[field] = [obj[field]];
+              }
+              obj[field].push(value);
+            } else {
+              obj[field] = value;
+            }
+            prev = field;
+          }
+        }
+      }
+
+      // Add last object if incomplete but has data
+      if (Object.keys(obj).length > 0) {
+        result.push({ ...context, ...obj });
+      }
+
+      return result;
+    }
+
+    // This modifies the parsed root
+    const pruned = prune(root);
```

Comment:
is the high level of this that it boils down the entire DOM tree into just the parts that match the selector?